File: packages/evidence/src/semver.ts
export interface SemVer {
  readonly major: number;
  readonly minor: number;
  readonly patch: number;
  readonly prerelease?: string;
}

export function parseSemVer(value: string): SemVer | null {
  const match = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?$/.exec(value);
  if (!match) return null;
  const base = { major: Number(match[1]), minor: Number(match[2]), patch: Number(match[3]) };
  return match[4] ? { ...base, prerelease: match[4] } : base;
}

export function compareSemVer(a: string, b: string): number {
  const left = parseSemVer(a);
  const right = parseSemVer(b);
  if (!left || !right) throw new Error(`Invalid semantic version: ${!left ? a : b}`);
  for (const key of ['major', 'minor', 'patch'] as const) {
    const delta = left[key] - right[key];
    if (delta !== 0) return delta;
  }
  if (left.prerelease === right.prerelease) return 0;
  if (!left.prerelease) return 1;
  if (!right.prerelease) return -1;
  return comparePrerelease(left.prerelease, right.prerelease);
}

export function satisfiesRange(version: string, range: string): boolean {
  const clauses = range.trim().split(/\s+/).filter(Boolean);
  if (!parseSemVer(version) || clauses.length === 0) return false;
  return clauses.every((clause) => {
    const match = /^(>=|<=|>|<|=)?((?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?)$/.exec(clause);
    if (!match) return false;
    const operator = match[1] ?? '=';
    const target = match[2] ?? '';
    const comparison = compareSemVer(version, target);
    return operator === '>=' ? comparison >= 0
      : operator === '<=' ? comparison <= 0
        : operator === '>' ? comparison > 0
          : operator === '<' ? comparison < 0
            : comparison === 0;
  });
}

function comparePrerelease(left: string, right: string): number {
  const a = left.split('.');
  const b = right.split('.');
  const count = Math.max(a.length, b.length);
  for (let index = 0; index < count; index += 1) {
    const leftPart = a[index];
    const rightPart = b[index];
    if (leftPart === undefined) return -1;
    if (rightPart === undefined) return 1;
    if (leftPart === rightPart) continue;
    const leftNumeric = /^\d+$/.test(leftPart);
    const rightNumeric = /^\d+$/.test(rightPart);
    if (leftNumeric && rightNumeric) return Number(leftPart) - Number(rightPart);
    if (leftNumeric) return -1;
    if (rightNumeric) return 1;
    return leftPart < rightPart ? -1 : 1;
  }
  return 0;
}

File: src/l9_assurance/evidence/semver.py
from __future__ import annotations

import re
from dataclasses import dataclass

_PATTERN = re.compile(
    r"^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?$"
)


@dataclass(frozen=True, slots=True)
class SemVer:
    major: int
    minor: int
    patch: int
    prerelease: str | None = None


def parse_semver(value: str) -> SemVer | None:
    match = _PATTERN.fullmatch(value)
    if match is None:
        return None
    return SemVer(int(match[1]), int(match[2]), int(match[3]), match[4])


def compare_semver(left_value: str, right_value: str) -> int:
    left = parse_semver(left_value)
    right = parse_semver(right_value)
    if left is None or right is None:
        raise ValueError(f"Invalid semantic version: {left_value if left is None else right_value}")
    for a, b in ((left.major, right.major), (left.minor, right.minor), (left.patch, right.patch)):
        if a != b:
            return a - b
    if left.prerelease == right.prerelease:
        return 0
    if left.prerelease is None:
        return 1
    if right.prerelease is None:
        return -1
    return _compare_prerelease(left.prerelease, right.prerelease)


def satisfies_range(version: str, expression: str) -> bool:
    if parse_semver(version) is None:
        return False
    clauses = [item for item in expression.strip().split() if item]
    if not clauses:
        return False
    for clause in clauses:
        match = re.fullmatch(r"(>=|<=|>|<|=)?(.+)", clause)
        if match is None or parse_semver(match[2]) is None:
            return False
        operator = match[1] or "="
        comparison = compare_semver(version, match[2])
        if not {
            ">=": comparison >= 0,
            "<=": comparison <= 0,
            ">": comparison > 0,
            "<": comparison < 0,
            "=": comparison == 0,
        }[operator]:
            return False
    return True


def _compare_prerelease(left: str, right: str) -> int:
    a = left.split(".")
    b = right.split(".")
    for index in range(max(len(a), len(b))):
        if index >= len(a):
            return -1
        if index >= len(b):
            return 1
        if a[index] == b[index]:
            continue
        left_numeric = a[index].isdigit()
        right_numeric = b[index].isdigit()
        if left_numeric and right_numeric:
            return int(a[index]) - int(b[index])
        if left_numeric:
            return -1
        if right_numeric:
            return 1
        return -1 if a[index] < b[index] else 1
    return 0
